### ISSUE TITLE
DCNC check not very flexible

### ISSUE THREAD

**Issue by jamiegau** [Ymagis/ClairMeta#151]
Hi,
Just getting into this tool for use with my workflows..
I find the dcnc check quite inflexible.  For example.  I small Indy film had some trailers for local distribution.
Using basic DCNC for the cpl..

The fact it errored on these basic naming conventions in a simply local DCP cpl name is a little broken if you ask me.

```
2020-07-29 23:45:33,504 - Clairmeta - INFO - Probing DCP : EMO_TLR_F_51_2K_20170418_FRO_IOP
2020-07-29 23:45:38,662 - Clairmeta - INFO - Total time : 5.16 seconds
2020-07-29 23:45:38,969 - Clairmeta - INFO - Checking DCP : EMO_TLR_F_51_2K_20170418_FRO_IOP
2020-07-29 23:45:52,463 - Clairmeta - INFO - DCP : EMO_TLR_F_51_2K_20170418_FRO_IOP
2020-07-29 23:45:52,463 - Clairmeta - INFO - Size : 908.61 MiB
2020-07-29 23:45:52,463 - Clairmeta - INFO - Warning(s) :
	check_dcnc_compliance - ContentTitle must have 12 parts, 8 found
ContentTitle Part 51 not matching any naming convention field
ContentTitle Part 2K not matching any naming convention field
ContentTitle Part 20170418 not matching any naming convention field
Field Language not found in ContentTitle
ContentTitle Part IOP not matching any naming convention field
2020-07-29 23:45:52,463 - Clairmeta - INFO - Total check : 59
2020-07-29 23:45:52,464 - Clairmeta - INFO - Total time : 13.48 sec
2020-07-29 23:45:52,464 - Clairmeta - INFO - Validation : Success

DCP - EMO_TLR_F_51_2K_20170418_FRO_IOP - Check succeeded
```
Is there any interest in actually improving this check to make it useful?  These false positives make it useless.  I will be disabling it in a profile I hope to create going forward...
I do like the idea of it helping to enforce better naming..  but this is not the answer.

**Comment by remia** [comment]
Hey @jamiegau,

What would you expect in this particular case ? Seems to me that this DCP is probably following an older version of the DCNC convention.

Would you expect the tool to check for every version in turn to find the best match ?

Errors messages could be improved with a bit more of complexity, here the most accurate is the one about the number of fields.

**Comment by jamiegau** [comment]
I have started to go tough the code to get a better understanding..

In general, many regional distributions do not follow the full 12 part parse "_" naming convention as many fields don't really make sense to follow when doing smaller releases.

Plus I found it very strange that obvious test like _51_ or _2K_ or _20200101_, obvious and unique in meaning could not be picked up with a more flexible parsing..  especially if the original 12 part parser failed to find them.
for example, if the date was not found do a regex for underscore, 8 numeric numbers, underscore. from section 4 onwards.  Or _2K_ will only every appear in the name because its 2K.  We can make some general assumptions like that.

I think it may be useful to add such attempts into the DCNC probe code.  Its likely not that difficult.

I may even give it a go.

**Comment by remia** [comment]
Yes the parser is not that great and the code is quite hard to read as well. I agree that things can certainly be improved, feel free to contribute if you feel you have a better way !

I tried a quick update here https://github.com/remia/ClairMeta/commit/6dc53956bf000a2503f7c81c2d2b210966f0fb5c , it should work better for your example (now reporting the correct missing 4 fields). Can't say if it'll work better for other configurations.

**Comment by jamiegau** [comment]
Thanks remia,
I'll have a look when I have time.
The DCDN has always been a "CONVENTION- a way in which something is usually done" meaning its flexible in its use.

It's interesting looking a the isdcf-DCNC website

`There MUST be an entry in EVERY field. Some systems count the number of “_” (underscores) to parse the information (NOT RECOMMENDED – BUT it’s done). If you have a non-relevant field, fill it with the letters “NULL”.`

It actually recommends that the method used in clairmeta is not recommended.
I do think, as this issue is marked, this should be considered a bug, a more flexible attempt is made to allow it to better deal with historical and CPL-names that are missing common sections.

Lets let this cook in the back of our minds for a while.  I more flexible implementation mya come to us.  If I have time I may fork and show you what I think would be appropriate.

**Comment by remia** [comment]
Sure, let me know when you have something to show.

To be honest I find this quote from ISDCF quite convoluted. Thinking about this once again, a way to further improve my proposal above would be to have a setting to mark some fields as optional, as in this case:
* Language
* TerritoryRating
* Studio
* PackageType

**Comment by jamiegau** [comment]
HI @remia 
I tried out your modification.  Working much better, though one issue..
I tried it to two generic DCP.
One:
```
python3 -m clairmeta.cli check -type dcp -progress /BigShare/UPLOAD/ClubMovieLOGOV1_ADV_F_51_2K_20140805
2020-08-19 10:38:01,241 - Clairmeta - INFO - Warning(s) :
	check_dcnc_compliance - CPL_6564340e-cf5c-4440-b5f4-0bf1b3a9c45c.xml : 
		ContentTitle must have 12 parts, 6 found
		ContentTitle Film Name does not respect naming convention rules : ClubMovieLOGOV1
		Field Language not found in ContentTitle
		Field TerritoryRating not found in ContentTitle
		Field Studio not found in ContentTitle
		Field Facility not found in ContentTitle
		Field Standard not found in ContentTitle
		Field PackageType not found in ContentTitle
	check_dcnc_field_claim_subtitle - CPL_6564340e-cf5c-4440-b5f4-0bf1b3a9c45c.xml : ContentTitle suggest Subtitle but CPL contains No Subtitle
```

and two:
```
python3 -m clairmeta.cli check -type dcp -progress /BigShare/UPLOAD/SCO-Austfilm_XSN_F_51_2K_20190611_SMPTE_OV/
2020-08-19 10:36:43,851 - Clairmeta - INFO - Error(s) :
	check_picture_cpl_encoding - CPL_2fef23c8-35c5-4d44-abcf-d6eebe43350d.xml (Asset jp2k_e046f07e-ae45-4262-93f8-f04087b5ced1_vid.mxf) : Picture must have 5 wavelet transform levels, 0 found
2020-08-19 10:36:43,851 - Clairmeta - INFO - Warning(s) :
	check_dcnc_compliance - CPL_2fef23c8-35c5-4d44-abcf-d6eebe43350d.xml : 
		ContentTitle must have 12 parts, 8 found
		Field Language not found in ContentTitle
		Field TerritoryRating not found in ContentTitle
		Field Studio not found in ContentTitle
		Field Facility not found in ContentTitle
	check_dcnc_field_claim_subtitle - CPL_2fef23c8-35c5-4d44-abcf-d6eebe43350d.xml : ContentTitle suggest Subtitle but CPL contains No Subtitle
```

One addition, where it prints or just under it...
`check_dcnc_compliance - CPL_2fef23c8-35c5-4d44-abcf-d6eebe43350d.xml :`
Add the actual CPL text the error is connected to.  When reviewing this error, you have no idea what the ContentTitleText actually is that its giving a warning against.

One ERROR:
`check_dcnc_field_claim_subtitle - CPL_2fef23c8-35c5-4d44-abcf-d6eebe43350d.xml : ContentTitle suggest Subtitle but CPL contains No Subtitle`
There is no suggestion that there is a Subtitle..  So its picking that up wrong.

That was a quick test on two DCP I had lying around that have very limited CPL names.
It would be great if could roll this into the release version, with the progress pull request soon.

**Comment by jamiegau** [comment]
One other comment,
`ContentTitle must have 12 parts, 6 found`
How about 
ContentTitle should have 12 parts, found 6, to be fully compliant with recommended practices.

**Comment by remia** [comment]
I tried to address your feedback in #153, could you give another try ?

**Comment by remia** [comment]
This was merged in #153
